Results from WPCLI that allows implement in w.org

davidperezgar opened the pull request:
We need to give a result from WPCLI that fits meta team in order to execute this plugin in form submission.

This results fits the need asked by Meta [#441 (comment)](https://github.com/WordPress/plugin-check/issues/441#issuecomment-2190961036), so we could implement as well as this #478 , and it would be the minimal needs to achieve that.

The results has a json format like this:
`{"pass":true,"errors":[],"warnings":[]}` 

The Pass value is true if it has no errors. 

I'd need some feedback if the approach is correct, or we could do something more.

davidperezgar committed:
alpha test for wporg form submission

davidperezgar committed:
better approach pass test

joemcgill commented:
As asked in https://github.com/WordPress/plugin-check/issues/441#issuecomment-2195224108

> I'm curious if the presence of any errors returned by the CLI command would be sufficient as a pass/fail signal, or if the command needs to also return a specific error code as well?

swissspidy commented on includes/CLI/Plugin_Check_Command.php:
```diff
@@ -98,6 +98,9 @@ public function __construct( Plugin_Context $plugin_context ) {
 	 * : Additional directories to exclude from checks
 	 * By default, `.git`, `vendor` and `node_modules` directories are excluded.
 	 *
+	 * [--wporg]
```
Instead of a new argument for this, it would make more sense to add another option to `--format`, e.g. `--format=wporg`

davidperezgar replied:
That was my first approach, but it gives more things related with that argument.

davidperezgar replied:
And We will for sure make changes in the plugin for needs from Meta, and It will easier for us keeping the command.

swissspidy replied:
I suppose long term, if you need more changes, you could also look into extending this class with your own custom command (e.g. `class Plugin_Review_Plugin_Check_Command extends Plugin_Check_Command`) or something.

Just from a code quality perspective a custom flag like this isn't exactly a best practice :)

davidperezgar replied:
done

davidperezgar commented:
> As asked in [#441 (comment)](https://github.com/WordPress/plugin-check/issues/441#issuecomment-2195224108)
> 
> > I'm curious if the presence of any errors returned by the CLI command would be sufficient as a pass/fail signal, or if the command needs to also return a specific error code as well?

I think is the easiest way. We have to be sure that errors haven't got any false positive.

davidperezgar committed:
passed argument wporg to format

mukeshpanchal27 commented on includes/CLI/Plugin_Check_Command.php:
```diff
@@ -77,6 +78,7 @@ public function __construct( Plugin_Context $plugin_context ) {
 	 *   - table
 	 *   - csv
 	 *   - json
+	 *   - wporg
```
Add new option in docblock.

`Format to display the results. Options are table, csv, json, and wporg. The default will be a table.`

mukeshpanchal27 commented on includes/CLI/Plugin_Check_Command.php:
```diff
@@ -210,6 +214,18 @@ static function ( $dirs ) use ( $excluded_directories ) {
 			$warnings = $result->get_warnings();
 		}
 
+		// Print the formatted results for WPORG.
+		if ( $run_wporg ) {
+			$all_results = array(
+				'pass'     => empty( $errors ) ? true : false,
+				'errors'   => $errors,
+				'warnings' => $warnings,
+			);
+
+			WP_CLI::line( json_encode( $all_results ) );
+			return;
+		}
```
```suggestion
		// Print the json formatted results for WPORG.
		if ( isset( $options['format'] ) && 'wporg' === $options['format'] ) {
			$wporg_results = array(
				'pass'     => empty( $errors ) ? true : false,
				'errors'   => $errors,
				'warnings' => $warnings,
			);

			WP_CLI::line( json_encode( $wporg_results ) );
			return;
		}
```

- Remove local variable and incorporate the condition inline.
- Slightly update the comment.

mukeshpanchal27 commented on includes/CLI/Plugin_Check_Command.php:
```diff
@@ -133,6 +135,8 @@ public function check( $args, $assoc_args ) {
 		$plugin = isset( $args[0] ) ? $args[0] : '';
 		$checks = wp_parse_list( $options['checks'] );
 
+		$run_wporg = isset( $options['format'] ) && 'wporg' === $options['format'] ? true : false;
+
 		// Create the categories array from CLI arguments.
```
```suggestion
		// Create the categories array from CLI arguments.
```

mukeshpanchal27 reviewed (changes requested):
Thanks @davidperezgar for the PR! Let some feedbacks other then it looks good to me.

- Fix `composer phpmd` errors.
   - Add `@SuppressWarnings(PHPMD.ExcessiveClassComplexity)` in class `Plugin_Check_Command` docblock.
   - Add `@SuppressWarnings(PHPMD.CyclomaticComplexity)` in `check` method docblck.
-  Additionally if you add Behat tests for new args then it would be great.

davidperezgar committed:
Update includes/CLI/Plugin_Check_Command.php

Co-authored-by: Mukesh Panchal <mukeshpanchal27@users.noreply.github.com>

davidperezgar committed:
Update includes/CLI/Plugin_Check_Command.php

Co-authored-by: Mukesh Panchal <mukeshpanchal27@users.noreply.github.com>

davidperezgar commented:
Thanks for your suggestions!

joemcgill commented:
I don't particularly like the approach of creating a new output format specific for wporg if we are instead able to consume one of the exiting formats when passing the correct arguments to the CLI script.

Before trying to implement a solution here, can we get an answer from systems (e.g., @dd32) to the question asked in https://github.com/WordPress/plugin-check/issues/441#issuecomment-2195224108? We also need to decide if #479 is a blocker to this issue, because we may not need a special CLI command at all if we can configure what errors are returned based on severity levels.

davidperezgar committed:
Merge branch 'trunk' into 441-plugin-review-team-output-for-wporg-form

davidperezgar committed:
Merge branch 'trunk' into 441-plugin-review-team-output-for-wporg-form

davidperezgar commented:
Related with #479 and #440 .

mukeshpanchal27 committed:
Merge branch 'trunk' into 441-plugin-review-team-output-for-wporg-form

mukeshpanchal27 committed:
Fix PHPMD report

mukeshpanchal27 committed:
Merge branch '441-plugin-review-team-output-for-wporg-form' of github.com:WordPress/plugin-check into 441-plugin-review-team-output-for-wporg-form

mukeshpanchal27 committed:
Add behat test

mukeshpanchal27 committed:
Update line no in behat test

mukeshpanchal27 commented:
@davidperezgar I have fixed PHPMD issues and added behat test for new argument.

ernilambar commented:
I tried this command:

```
wp plugin check foo-single.php --fields=line,column,code --format=wporg
```

Argument `--fields` is not respected when using `wporg` format.

davidperezgar commented:
Yes, and it's correct. The format for wporg is very strict and I shouldn't be customized.

dd32 commented:
I don't see anything "wrong" here, but I don't see the purpose of it, other than to fix the fact that `--format=json` isn't a pure JSON output.

Instead of adding this for the purposes of adding it to w.org plugin submission, I'd probably just suggest fixing the existing json output, or request that we parse the existing json output.

My main issue with this PR is that:
> 'pass'     => empty( $errors ) ? true : false,

Is completely useless. Pretty much every plugin throws up errors in plugin-check that are considered errors that aren't actually errors.

The original "A command that can be run, parsed, and provides a clear block or accept result" really still stands, and while this does do that, it still doesn't provide a the last part in any form that's inline with the plugin submission guidelines IMHO.

For example; Are this really something that should block a plugin submission?
> | 0    | 0      | ERROR   | outdated_tested_upto_header | Tested up to: 6.5 < 6.6                                                                     |
> | 37   | 25     | ERROR | WordPress.WP.I18n.MissingTranslatorsComment | A function call to `__()` with texts containing placeholders was found, but was not accompanied by a "translators:" comment on the line above to clarify the meaning of the placeholders. |
> | 59   | 21     | ERROR | WordPress.Security.EscapeOutput.ExceptionNotEscaped | All output should be run through an escaping function (see the Security sections in the WordPress Developer Handbooks), found `'__'`.            |
> | 114  | 30     | ERROR   | WordPress.WP.AlternativeFunctions.json_encode_json_encode | json_encode() is discouraged. Use wp_json_encode() instead. |

So I guess actually, this all comes back to #479 instead as the main blocker.

davidperezgar commented:
Yes, We're working on severity levels, and the team where agreed equals and more than 7, should be an error. 

So, I see:
- We should have a format=wporg as json is for users.
- We have we give you then after the severity level an ERROR or WARNING
- Do you need then the pass variable?

For the example that you give: 
1. ERROR
2. WARNING
3. ERROR
4. ERROR

So we have to work on severity levels and then give you the result.

davidperezgar committed:
Merge branch 'trunk' into 441-plugin-review-team-output-for-wporg-form

davidperezgar committed:
changes for wporg and severity

davidperezgar commented:
Hello, 
I've updated the code with the severity level and wporg format. I've made within iterative for all formats, but creating an array for errors and warning after severity filter.

I've removed pass as is useless.

We can discuss what's your opinion.

davidperezgar committed:
fix phplint

ernilambar committed:
Update format output and add feature tests

ernilambar committed:
Update command docs

ernilambar committed:
Fix missing error results

ernilambar committed:
Avoid time dependent check in feature tests

davidperezgar commented:
Thanks Nilambar for the changes, I think is ready for a review from @dd32

dd32 commented:
Am I understanding this correctly?

```
$output = json_decode( wp plugin-check .... )
$error_types = wp_list_pluck( $output->errors, 'type' );
$pass = ! in_array( 'ERROR', $error_types );
```

If so, yep, that's fine.

davidperezgar commented:
Yes Dion! ERRORS will block an upload. But please show as well warnings as they will help to reduce the time for reviewing.

swissspidy commented on includes/CLI/Plugin_Check_Command.php:
```diff
@@ -248,6 +250,15 @@ static function ( $dirs ) use ( $excluded_files ) {
 		$error_severity   = ! empty( $options['error-severity'] ) ? $options['error-severity'] : $options['severity'];
 		$warning_severity = ! empty( $options['warning-severity'] ) ? $options['warning-severity'] : $options['severity'];
 
+		$format_is_wporg = ( isset( $options['format'] ) && 'wporg' === $options['format'] ) ? true : false;
```
```suggestion
		$format_is_wporg = ( isset( $options['format'] ) && 'wporg' === $options['format'] );
```

ernilambar committed:
Improve conditional for wporg

Co-authored-by: Pascal Birchler <pascalb@google.com>

ernilambar committed:
Merge branch 'trunk' into 441-plugin-review-team-output-for-wporg-form

ernilambar committed:
Update message strings which were recently changed

davidperezgar committed:
Merge branch 'trunk' into 441-plugin-review-team-output-for-wporg-form

davidperezgar commented:
@dd32 Do you need something else about this PR? Thanks.

ernilambar committed:
Merge branch 'trunk' into 441-plugin-review-team-output-for-wporg-form

dd32 reviewed (approved):
> @dd32 Do you need something else about this PR? Thanks.

Nope, It was just waiting on merging the other needed PRs AFAIK.

dd32 committed:
Merge branch 'trunk' into 441-plugin-review-team-output-for-wporg-form

mukeshpanchal27 commented on includes/CLI/Plugin_Check_Command.php:
```diff
@@ -248,6 +250,15 @@ static function ( $dirs ) use ( $excluded_files ) {
 		$error_severity   = ! empty( $options['error-severity'] ) ? $options['error-severity'] : $options['severity'];
 		$warning_severity = ! empty( $options['warning-severity'] ) ? $options['warning-severity'] : $options['severity'];
 
+		$format_is_wporg = ( isset( $options['format'] ) && 'wporg' === $options['format'] );
```
```suggestion
		$is_wporg_format = isset( $options['format'] ) && 'wporg' === $options['format'];
```

Nit. name should be `is_wporg_format`

davidperezgar replied:
ok!

mukeshpanchal27 commented on includes/CLI/Plugin_Check_Command.php:
```diff
@@ -676,4 +700,33 @@ function ( $item ) use ( $warning_severity ) {
+	/**
+	 * Returns separated results by ERROR and WARNING.
+	 *
+	 * @since 1.1.0
```
```suggestion
	 * @since 1.2.0
```

Should be upcoming version.

mukeshpanchal27 commented on includes/CLI/Plugin_Check_Command.php:
```diff
@@ -676,4 +700,33 @@ function ( $item ) use ( $warning_severity ) {
+	 * @param array $results Check results.
+	 * @return array Separated results.
+	 */
+	private function get_separated_results( $results ) {
+		$errors = array_filter(
+			$results,
+			function ( $item ) {
+				return ( 'ERROR' === $item['type'] );
+			}
+		);
+
+		$warnings = array_filter(
+			$results,
+			function ( $item ) {
+				return ( 'WARNING' === $item['type'] );
+			}
+		);
+
+		return array(
+			'errors'   => array_values( $errors ),
+			'warnings' => array_values( $warnings ),
+		);
+	}
```
```suggestion
	private function get_separated_results( $results ) {
		// Initialize arrays for errors and warnings.
		$errors   = array();
		$warnings = array();

		// Return early if $results is not an array or is empty.
		if ( ! is_array( $results ) || empty( $results ) ) {
			return array(
				'errors'   => $errors,
				'warnings' => $warnings,
			);
		}

		// Separate errors and warnings in one pass.
		foreach ( $results as $item ) {
			if ( ! isset( $item['type'] ) ) {
				continue;
			}

			switch ( $item['type'] ) {
				case 'ERROR':
					$errors[] = $item;
					break;
				case 'WARNING':
					$warnings[] = $item;
					break;
			}
		}

		return array(
			'errors'   => $errors,
			'warnings' => $warnings,
		);
	}
```

- Early return if no error.
- Refactor code.

dd32 reviewed (changes requested):
Let's not merge this, until it's determined to be needed. See my comment above on #441

davidperezgar committed:
fixes variable name

davidperezgar committed:
suggested changes from Mukesh

davidperezgar committed:
Merge branch 'trunk' into 441-plugin-review-team-output-for-wporg-form

ernilambar commented:
Probably we dont need this PR now. It seems it is possible to integrate in org without new format.

dd32 commented:
Closing, we've taken a different direction in implementation. See https://meta.trac.wordpress.org/ticket/7778

davidperezgar commented:
Ok! No problem, if the result is the same! ;)
